File: LeetCode/src/greedy/JumpGame.java
package greedy;

public class JumpGame {
    public boolean canJump(int[] nums) {
        /*
        int farthest = 0;

        for (int i = 0; i < nums.length; i++) {
            if (i > farthest) {
                return false;
            }
            farthest = Math.max(farthest, i + nums[i]);
        }

        return true;
         */
        
        int last = nums.length - 1;

        for (int i = nums.length - 1; i >= 0; i--) {
            if (i + nums[i] >= last) {
                last = i;
            }
        }

        return last == 0;
    }

    public static void main(String[] args) {
        JumpGame solution = new JumpGame();

        int[] test1 = {2, 3, 1, 1, 4};
        System.out.println(solution.canJump(test1)); // Output: true

        int[] test2 = {3, 2, 1, 0, 4};
        System.out.println(solution.canJump(test2)); // Output: false
    }
}

File: LeetCode_Python/greedy/JumpGame.py
class JumpGame:
    def canJump(self, nums: list[int]) -> bool:
        n = len(nums)
        fastest = 0

        for i in range(n):
            if i > fastest:
                return False
            fastest = max(fastest, i + nums[i])
            if fastest >= n - 1:
                return True
        return False


# Time Complexity: O(n)
# Space Complexity: O(1)

if __name__ == "__main__":
    solution = JumpGame()

    print("Expeted Output : True")
    print("Actual Output : ", solution.canJump([2, 3, 1, 1, 4]))

    print("Expeted Output : False")
    print("Actual Output : ", solution.canJump([3, 2, 1, 0, 4]))
